### PR TITLE
Gatv2 

### PR DESCRIPTION
**Pull request by yxnyu**
## Description
This PR integrates Graph Attention Network (GAT) model into the OpenADMET anvil workflow system, enabling users to run GAT-based molecular property prediction through the standard `openadmet anvil --recipe-path` command interface.

The integration includes a complete GATv2 implementation with PyTorch Lightning, molecular graph featurization using RDKit, and seamless integration with the existing OpenADMET workflow architecture.

## Todos
Notable points that this PR has accomplished:
  - [x] Implement GATv2ModelWrapper with PyTorch Lightning integration
  - [x] Add GATGraphFeaturizer for molecular graph featurization using RDKit
  - [x] Register GAT model and featurizer in OpenADMET registries
  - [x] Modify AnvilDeepLearningWorkflow to support PyTorch Geometric DataLoaders
  - [x] Update LightningTrainer to handle optional validation dataloaders
  - [x] Add proper data splitting (train/validation/test) for deep learning workflow
  - [x] Create basic_anvil_gat.yaml configuration example
  - [x] Add automatic input/edge dimension inference from training data
  - [x] Update .gitignore to exclude temporary lightning logs
  - [x] Test full workflow integration with CYP3A4 dataset

## Questions
- [x] Should GAT models use CPU or GPU by default? (Resolved: Default to CPU for compatibility)
- [x] How to handle MPS GPU compatibility issues on Mac? (Resolved: Added fallback mechanism)
- [x] Should validation set be mandatory for GAT training? (Resolved: Yes, implemented 70/10/20 split)

## Status
- [x] Ready to go

## Testing
Tested with CYP3A4 dataset using the command:
```bash
python -m openadmet anvil --recipe-path openadmet/models/tests/test_data/basic_anvil_gat.yaml
```

### CODE REVIEW

File: openadmet/models/architecture/gat.py [hunk 1]
```diff
@@ -0,0 +1,650 @@
...
+            "num_attention_heads": self.num_heads,
+            "hidden_dimension": self.hidden_dim,
+            "pooling_method": self.pooling,
+            "dropout_rate": self.gat_dropout
+        }
+
+        return summary
+
+
+def test_gat_from_yaml():
```
Comment (hmacdope) [inline openadmet/models/architecture/gat.py:466]: Move this to the test area `openadmet/models/tests/anvil`. You can also add files to help with testing.

File: openadmet/models/architecture/gat.py [hunk 2]
```diff
@@ -0,0 +1,650 @@
...
+    #       ...
+
+    if "procedure" not in config or "model" not in config["procedure"] or "params" not in config["procedure"]["model"]:
+        raise ValueError("YAML file must contain procedure.model.params section")
+
+    wrapper_params = config["procedure"]["model"].get("params", {})
+    return GATv2ModelWrapper.from_params(class_params=wrapper_params, model_params=model_params)
+
+
+def run_workflow_from_config(config_path: str):
```
Comment (hmacdope) [inline openadmet/models/architecture/gat.py:519]: Same from here, remove, should be doable from Anvil.

File: openadmet/models/features/gat_featurizer.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
+from typing import Optional, List, Any
+import torch
+from torch_geometric.data import Data
+from torch_geometric.loader import DataLoader
+from loguru import logger
+
+try:
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:7]: Should just be an error if no `rdkit`. I wouldn't even bother with this guard, just let it fail with `ImportError` for rdkit.

File: openadmet/models/features/gat_featurizer.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+    from rdkit import Chem
+except ImportError:
+
+    logger.warning(
+        "RDKit library not found. GraphFeaturizer will not be functional. "
+        "Please install RDKit (e.g., conda install -c conda-forge rdkit)."
+    )
+    Chem = None # Set to None to check its availability later
+
+# Assuming feature_base.py is in the same directory or accessible via PYTHONPATH
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:17]: Is there a specific reason this should be necessary? Registering and also importing in `openadmet/models/registries.py` should do the job.

File: openadmet/models/features/gat_featurizer.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    """
+    Featurizer to convert SMILES strings into graph Data objects suitable for GAT-like models.
+    It extracts atom features and bond features using RDKit.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self._prepare()
+
+    def _prepare(self):
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:47]: Remove this, assume user has right environment installed, unless there is something I am missing.

File: openadmet/models/features/gat_featurizer.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        Check if RDKit is available.
+        """
+        if Chem is None:
+            raise ImportError(
+                "RDKit library is not installed, which is essential for GATGraphFeaturizer. "
+                "Please install RDKit."
+            )
+        logger.info("GATGraphFeaturizer initialized with RDKit.")
+
+    def _featurize_single_molecule(self, smiles: str, y_val: Optional[float] = None) -> Optional[Data]:
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:58]: @kaminow would you be able to take a look over the featurization here?

File: openadmet/models/tests/test_data/basic_anvil_gat.yaml [hunk 7]
```diff
@@ -0,0 +1,68 @@
+metadata:
```
Comment (hmacdope) [inline openadmet/models/tests/test_data/basic_anvil_gat.yaml:1]: Amazing!

File: openadmet/models/trainer/lightning.py.bak [hunk 8]
```diff
@@ -0,0 +1,121 @@
+from pathlib import Path
```
Comment (hmacdope) [inline openadmet/models/trainer/lightning.py.bak:1]: Remove this file

File: openadmet/models/features/gat_featurizer.py [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+                data_objects.append(data)
+
+        # Create DataLoader
+        dataloader = DataLoader(
+            data_objects,
+            batch_size=batch_size,
+            shuffle=shuffle,
+            num_workers=num_workers
+        )
+
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:173]: Should also return an element in second place that lists the indices of the molecules that were correctly featurized. See the example here: https://github.com/OpenADMET/openadmet-models/pull/186/files#diff-57f3450abffd4b8ff6f31648a86b47e79ba78b11230bbae4251b6c72f952dee1 from #186 which will need to go in first. 

I.E the API becomes `dataloader, sucessfull_idx, scaler, objs`

File: openadmet/models/features/gat_featurizer.py [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+            edge_attr = None # No edges, so no edge attributes
+
+        data = Data(x=x, edge_index=edge_index, edge_attr=edge_attr)
+
+        if y_val is not None:
+            data.y = torch.tensor([y_val], dtype=torch.float)
+
+        return data
+
+    def featurize(self, smiles_list: list[str], y_list: Optional[list[Any]] = None, batch_size: int = 32, shuffle: bool = False, num_workers: int = 0):
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:117]: Use the same API names as the ChemProp Featurizer, e.g `smiles_list -> smiles` and `y_list` -> `y`

You should also make `batch_size`, `shuffle` and `num_workers` class attributes rather than passed to `featurize`

The signature from ChemProp featurizer is shown below for ease. 
```python
    def featurize(
        self, smiles: Iterable[str], y: Iterable[Any] = None
    ) -> tuple[
        DataLoader,
        np.ndarray,
        StandardScaler,
        Union[MoleculeDataset, ReactionDataset, MulticomponentDataset],
    ]:
```

File: openadmet/models/features/gat_featurizer.py [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+        data_objects = []
+        # Convert y_list to a list if it's a pandas Series to avoid indexing issues
+        if y_list is not None:
+            if hasattr(y_list, 'tolist'):
+                y_list = y_list.tolist()
+            elif hasattr(y_list, 'values'):
+                y_list = y_list.values.tolist()
+
+        for i, smiles_str in enumerate(smiles_list):
```
Comment (hmacdope) [inline openadmet/models/features/gat_featurizer.py:145]: I wouldn't bother with this, expect people to put in numerical values or thrown an error.

File: openadmet/models/anvil/workflow.py [hunk 12]
```diff
@@ -585,7 +585,13 @@ def run(
 
         # Build model
         logger.info("Building model")
-        self.model.build(scaler=train_scaler)
+        # Pass train_dataloader to build method for models that need to infer dimensions
+        if hasattr(self.model, 'build'):
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/anvil/workflow.py:590]: I think it should affect some auto-dealing module but it is ok, I fixed it

File: openadmet/models/architecture/gat.py [hunk 13]
```diff
@@ -0,0 +1,650 @@
...
+
+        self.model = GATv2Model(**model_config)
+        self.loss_fn = self._get_loss_function(loss_fn_name)
+        self.lr = lr
+        self.weight_decay = weight_decay
+        self.scheduler_name = scheduler_name
+        self.warmup_epochs = warmup_epochs
+
+    def _get_loss_function(self, name: str):
+        loss_functions = {
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/architecture/gat.py:186]: fixed!

File: openadmet/models/architecture/gat.py [hunk 14]
```diff
@@ -0,0 +1,650 @@
...
+            if residual is not None and residual.size(-1) == x.size(-1):
+                x = x + residual
+
+        # Graph-level pooling
+        x = self.pool(x, batch)
+
+        # Output layers
+        out = self.output_layers(x)
+
+        return out
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/architecture/gat.py:159]: thx!

File: openadmet/models/features/gat_featurizer.py (hunk 10 above)
Comment (yxnyu) [reply openadmet/models/features/gat_featurizer.py:117]: got it!

File: openadmet/models/tests/test_data/basic_anvil_gat.yaml (hunk 7 above)
Comment (yxnyu) [reply openadmet/models/tests/test_data/basic_anvil_gat.yaml:1]: thx

**Comment by yxnyu** [comment]
the latest version is updated

**Comment by yxnyu** [comment]
update d

**Review by smcolby — CHANGES_REQUESTED** [review]
Review ready when you are @yxnyu!

File: openadmet/models/tests/test_data/basic_anvil_gat.yaml [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  target_cols:
+    - pChEMBL mean
+
+procedure:
+  split:
+    type: "ShuffleSplitter"
+    params:
+      val_size: 0.1
+      test_size: 0.1
+      shuffle: True
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/tests/test_data/basic_anvil_gat.yaml:26]: In /openadmet-models/openadmet/models/split/split_base.py, there is saying if       
  if self.test_size + self.val_size + self.train_size != 1.0:
            raise ValueError("Test and train sizes must sum to 1.0")
I think you‘re right and I  explicit with train_size=0.8

File: openadmet/models/anvil/data_spec.py [hunk 16]
```diff
@@ -66,6 +66,8 @@ def read(self) -> tuple[pd.Series, pd.Series]:
             raise ValueError(f"Unsupported resource type: {self.resource}")
         # now read the target columns and smiles column
         targets = data[self.target_cols]
+        if isinstance(self.target_cols, list) and len(self.target_cols) == 1:
+            targets = targets[self.target_cols[0]]
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/anvil/data_spec.py:70]: done, we fix it in the gat_feature.py

File: openadmet/models/architecture/gat.py [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+
+        if self.scheduler_name == "cosine":
+            scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+                optimizer, T_max=self.trainer.max_epochs if self.trainer else 100
+            )
+            return [optimizer], [{"scheduler": scheduler, "interval": "epoch"}]
+        elif self.scheduler_name == "reduce_on_plateau":
+            scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+                optimizer, mode='min', factor=0.5, patience=10
+            )
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/architecture/gat.py:252]: changed it!

File: openadmet/models/architecture/gat.py [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+
+    try:
+        script_dir = os.path.dirname(os.path.abspath(__file__))
+        default_yaml_path = os.path.join(script_dir, '..', 'tests', 'test_data', 'basic_anvil_gat.yaml')
+        # Since run_workflow_from_config is removed, this part will be cleaned up.
+        # Placeholder for potential future main-guard use.
+    except Exception as e:
+        import traceback
+        logger.error(f"An error occurred: {e}")
+        logger.error(f"Full Traceback:\n{traceback.format_exc()}")
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/architecture/gat.py:458]: removed!

File: openadmet/models/architecture/gat.py [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+                x = F.relu(x)
+                x = F.dropout(x, p=self.dropout, training=self.training)
+
+            # Residual connection (if dimensions match)
+            if residual is not None and residual.size(-1) == x.size(-1):
+                x = x + residual
+
+        # Graph-level pooling
+        x = self.pool(x, batch)
+
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/architecture/gat.py:152]: I removed the redundant code

File: openadmet/models/features/gat_featurizer.py [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+        # Create DataLoader
+        dataloader = DataLoader(
+            data_objects,
+            batch_size=self.batch_size,
+            shuffle=self.shuffle,
+            num_workers=self.num_workers,
+        )
+
+        # Return dataloader, indices, scaler (None for GAT), and dataset (data_objects)
+        return dataloader, np.array(successful_indices), None, data_objects
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/features/gat_featurizer.py:146]: None indicates that the features generated by GAT do not require normalization or scaling.

[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/features/gat_featurizer.py:146]: as to Training hyperparameters should indeed be separate and part of the trainer implementation., I try but seems more ugly?

File: openadmet/models/tests/models/test_base.py [hunk 21]
```diff
@@ -1,10 +1,15 @@
...
 
-from openadmet.models.architecture.model_base import models
+from openadmet.models.architecture.model_base import (
+    PickleableModelBase,
+    models,
+)
 
 
-@pytest.mark.parametrize("mclass", models.classes())
+@pytest.mark.parametrize("mclass", models.values())
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/tests/models/test_base.py:9]: i think a values is better but I return into the original one.

File: openadmet/models/tests/models/test_base.py [hunk 22]
```diff
@@ -1,10 +1,15 @@
...
 def test_save_load_pickleable(mclass, tmp_path):
+    if not issubclass(mclass, PickleableModelBase):
+        pytest.skip(f"Skipping non-pickleable model {mclass.__name__}")
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/tests/models/test_base.py:12]: PickleableModelBase is a Scikit-learn–style model. GAT is torch based. but I’ve introduced a test suite for TorchModelBase to verify that models support both pickle-based and torch-based serialization and deserialization.

File: openadmet/models/trainer/lightning.py [hunk 23]
```diff
@@ -105,7 +105,7 @@ def prepare(self):
             accelerator=self.accelerator,
             devices=self.devices,  # Use GPU if available
             max_epochs=self.max_epochs,  # number of epochs to train for
-            callbacks=[checkpointing],  # Use the configured checkpoint callback
+            callbacks=self._callbacks,  # Use the configured checkpoint callback
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/trainer/lightning.py]: fixed!

File: openadmet/models/tests/test_data/basic_anvil_gat.yaml [hunk 24]
```diff
@@ -0,0 +1,61 @@
...
+      hidden_dim: 64
+      num_layers: 4
+      num_heads: 8
+      gat_dropout: 0.1
+      pooling: "mean"
+      output_dim: 1
+      loss_function: "mse"
+
+  train:
+    type: "LightningTrainer"
```
[the remark this answers is not in the document]
Comment (yxnyu) [reply openadmet/models/tests/test_data/basic_anvil_gat.yaml]: I had make a new Trainer
